collections/_posts/2019-08-24-adding-custom-doh-resolvers-to-dnscloak.md

```diff
@@ -0,0 +1,88 @@
+---
+title: Adding Custom DNS over HTTPS Resolvers to DNSCloak
+author: nitrohorse
+layout: post
+tags:
+  - guides
+  - dns
+  - ios
+---
+
+**[DNSCloak](https://apps.apple.com/us/app/dnscloak-secure-dns-client/id1452162351)** is an [open-source](https://github.com/s-s/dnscloak) DNSCrypt and DNS over HTTPS (DoH) client for iOS; giving users the ability to encrypt their DNS requests. While highly configurable, its user interface can be unintuitive and doesn't easily allow users to add custom DoH resolvers apart from the default ["public-resolvers" list that DNSCrypt provides](https://github.com/DNSCrypt/dnscrypt-resolvers/blob/master/v2/public-resolvers.md).
+
+This guide will show how to use DNSCloak to connect to your favorite public resolver that supports DNSCrypt or DoH.
+
+### Adding a Custom Resolver
+
+You'll notice DNSCloak provides a "Config Editor":
+
+![](/assets/img/2019-08-24-dnscloak/config-editor.jpeg){: .w-50}
+
+We can learn about the various configuration options from the [example config file](https://github.com/jedisct1/dnscrypt-proxy/blob/master/dnscrypt-proxy/example-dnscrypt-proxy.toml) in dnscrypt-proxy's code repository. If you scroll all the way to the bottom, you'll find a `[static.'myserver']` section along with a `stamp` property. This stamp is for adding the [DNS stamp](https://dnscrypt.info/stamps-specifications) of your resolver.
+
+> Server stamps encode all the parameters required to connect to a secure DNS server as a single string. Think about stamps as QR code, but for DNS.
+
+At the time of writing this post, CZ.NIC's DoH resolver is the only one [PrivacyTools.io suggests](https://www.privacytools.io/providers/dns/#icanndns) that doesn't provide their users with a DNS stamp on their website, making adoption a bit more difficult. But thankfully, we can create a DNS stamp ourselves.
+
+DNSCrypt hosts a [DNS stamp calculator](https://dnscrypt.info/stamps/) (which you can also [download, compile, and run offline](https://github.com/jedisct1/vue-dnsstamp)) that we can fill out with the info CZ.NIC provides on their webpage to generate our stamp.
+
+We'll need three things about the DoH resolver:
```

Comment:
```suggestion
We will need to know three things about the DoH resolver you choose:
```

DoH only? This should also work with DNSCrypt right?

Reply:
Yeah, I'm specifically walking through DoH resolvers because I haven't personally done this with DNSCrypt resolvers -- the stamp requirements are also different in that they need:
- IP address
- "Provider public key"
- and a "Provider name"